components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/AbstractComplexController.java

```diff
@@ -287,26 +306,25 @@ public void save(Patient patient)
             XWikiDocument doc = (XWikiDocument) this.documentAccessBridge.getDocument(patient.getDocument());
             BaseObject dataHolder = doc.getXObject(getXClassReference());
             PatientData<T> data = patient.getData(this.getName());
-            if (dataHolder == null && data != null) {
+            if (dataHolder == null || data == null) {
                 return;
             }
             XWikiContext context = this.contextProvider.get();
             for (String propertyName : getProperties()) {
                 Object propertyValue = data.get(propertyName);
-                if (propertyValue == null) {
-                    continue;
-                }
-                if (this.getCodeFields().contains(propertyName) && this.isCodeFieldsOnly()) {
-                    @SuppressWarnings("unchecked")
-                    List<VocabularyProperty> terms = (List<VocabularyProperty>) propertyValue;
-                    List<String> listToStore = new LinkedList<>();
-                    for (VocabularyProperty term : terms) {
-                        String name = StringUtils.isNotBlank(term.getId()) ? term.getId() : term.getName();
-                        listToStore.add(name);
+                if (isInKeySet(data, propertyName)) {
```

Comment:
after PT-2670 is merged in, we can just write if (data.hasKey(propertyName)) .... and remove isInKeySet()